lib/active_merchant/billing/gateways/trexle.rb

```diff
@@ -0,0 +1,219 @@
+          gsub(/(number\\?":\\?")(\d*)/, '\1[FILTERED]').
+          gsub(/(cvc\\?":\\?")(\d*)/, '\1[FILTERED]')
+      end
+      private
+
+      def add_amount(post, money, options)
+        post[:amount] = amount(money)
+        post[:currency] = (options[:currency] || currency(money))
+        post[:currency] = post[:currency].upcase if post[:currency]
+      end
+
+      def add_customer_data(post, options)
+        post[:email] = options[:email] if options[:email]
+        post[:ip_address] = options[:ip] if options[:ip]
+      end
+
+      def add_address(post, creditcard, options)
+        return if creditcard.kind_of?(String)
+        address = (options[:billing_address] || options[:address])
+        return unless address
+
+        post[:card] ||= {}
+        post[:card].merge!(
+          :address_line1 => address[:address1],
+          :address_line2 => address[:address_line2],
+          :address_city => address[:city],
+          :address_postcode => address[:zip],
+          :address_state => address[:state],
+          :address_country => address[:country]
+        )
+      end
+
+      def add_invoice(post, options)
+        post[:description] = options[:description] || "Active Merchant Purchase"
+      end
+
+      def add_capture(post, options)
+        capture = options[:capture]
+
+        post[:capture] = capture == false ? false : true
+      end
+
+      def add_creditcard(post, creditcard)
+        if creditcard.respond_to?(:number)
+          post[:card] ||= {}
+
+          post[:card].merge!(
+            :number => creditcard.number,
+            :expiry_month => creditcard.month,
+            :expiry_year => creditcard.year,
+            :cvc => creditcard.verification_value,
+            :name => creditcard.name
+          )
+        elsif creditcard.kind_of?(String)
+          if creditcard =~ /^token_/
+            post[:card_token] = creditcard
+          else
+            post[:customer_token] = creditcard
+          end
+        end
+      end
+
+      def headers(params = {})
+        result = {
+          "Content-Type" => "application/json",
+          "Authorization" => "Basic #{Base64.strict_encode64(options[:api_key] + ':').strip}"
+        }
+
+        result['X-Partner-Key'] = params[:partner_key] if params[:partner_key]
+        result['X-Safe-Card'] = params[:safe_card] if params[:safe_card]
+        result
+      end
+
+      def commit(method, action, params, options)
+        url = "#{test? ? test_url : live_url}/#{action}"
+
+        begin
+          raw_response = ssl_request(method, url, post_data(params), headers(options))
+          body = parse(raw_response)
+        rescue ResponseError => e
+          body = parse(e.response.body)
+        end
+
+        if body["response"]
+          success_response(body)
+        elsif body["error"]
+          error_response(body)
+        end
+
+      rescue JSON::ParserError
+        return unparsable_response(raw_response)
+      end
```

Comment:
As you have it implemented right now, every caller to `#success_response` would have to know to check for `body["response"]` first, as you have it now.

The smell is that I could theoretically write:
```ruby
body = { 'invalid_key': 'invalid_value' }
success_response(body) 
```
with possibly unexpected results, because `#success_response` just assumes that `body["response"]` is there. What if it isn't? 😉 

So that's why I'd prefer to see: 
```ruby
def success_response(body) 
  return uh_oh_something_bad_happened unless body['response']

  # do things now that I've ensured I have a body that I can work with
end
```
and also why I posed the questions of what to do if these functions are called with `nil` or a Hash that's not formed as you expect. (This can happen if suddenly the response schema from a gateway changes even slightly!)